src/stacks-hierarchy/saved-payment-utils.ts

```diff
@@ -1,51 +1,81 @@
-import {RecurringPayment} from '@atb/ticketing';
+import {PaymentMethod} from './types';
+import {useFirestoreConfiguration} from '@atb/configuration';
+import {parseISO} from 'date-fns';
 
 export function usePreviousPaymentMethods(): {
-  recurringPayments: RecurringPayment[] | undefined;
-  previousPaymentMethod: SavedPaymentOption | undefined;
+  recurringPaymentMethods: PaymentMethod[] | undefined;
+  previousPaymentMethod: PaymentMethod | undefined;
 } {
   const {userId} = useAuthState();
   const {data: recurringPayments} = useListRecurringPaymentsQuery();
   const [previousPaymentMethod, setPreviousPaymentMethod] =
-    useState<SavedPaymentOption>();
+    useState<PaymentMethod>();
+  const {paymentTypes} = useFirestoreConfiguration();
+
+  const isValidPaymentMethod = useCallback(
+    (paymentMethod: PaymentMethod | undefined) => {
+      if (!paymentMethod) return false;
+
+      // Since the stored payment could have been deleted, we need to check if
+      // it is still in the list of recurring payments.
+      if (paymentMethod.recurringCard) {
+        const recurringPayment = recurringPayments?.find(
+          (recurringPayment) =>
+            recurringPayment.id === paymentMethod.recurringCard?.id,
+        );
+        if (!recurringPayment) return false;
+      }
+
+      // Payment type is not enabled
+      if (!paymentTypes.includes(paymentMethod.paymentType)) return false;
+
+      // Card has expired
+      const expired =
+        paymentMethod.recurringCard &&
+        parseISO(paymentMethod.recurringCard.expires_at).getTime() < Date.now();
+      if (expired) return false;
+
+      return true;
+    },
+    [recurringPayments, paymentTypes],
+  );
 
   useEffect(() => {
     if (!userId) {
       setPreviousPaymentMethod(undefined);
-    } else {
-      getPreviousPaymentMethodByUser(userId).then((storedMethod) => {
-        // Since the stored payment could have been deleted, we need to check if
-        // it is still in the list of recurring payments.
-        if (!storedMethod) return;
-        const shouldUseStored =
-          'recurringCard' in storedMethod
-            ? !!recurringPayments?.find(
-                (recurringPayment) =>
-                  recurringPayment.id === storedMethod.recurringCard.id,
-              )
-            : true;
-        setPreviousPaymentMethod(shouldUseStored ? storedMethod : undefined);
-      });
+      return;
     }
-  }, [userId, recurringPayments]);
+    getPreviousPaymentMethodByUser(userId).then((storedMethod) => {
+      setPreviousPaymentMethod(
+        isValidPaymentMethod(storedMethod) ? storedMethod : undefined,
+      );
+    });
+  }, [userId, isValidPaymentMethod]);
+
+  const recurringPaymentMethods = recurringPayments?.map(
+    (recurringPayment): PaymentMethod => ({
+      savedType: 'recurring',
+      paymentType: recurringPayment.payment_type,
+      recurringCard: recurringPayment,
```

Comment:
While we're at it - should we do something about this one as well?

<img width="251" alt="Skjermbilde 2024-09-06 kl  13 04 27" src="https://github.com/user-attachments/assets/1faea6fa-83db-4869-9ca6-9d6e5806084a">
<img width="273" alt="Skjermbilde 2024-09-06 kl  13 04 44" src="https://github.com/user-attachments/assets/6489e845-c46a-4368-a23a-be76d5c463a7">

Reply:
Yes, nice! Getting rid of RecurringCard 🔥